Name: Shuyun Wen

Email:wendywen027@gmail.com

#NOTE:
Please insert a text cell at the top of the Notebook with your Full Name in Bold and your email that you use for both Journey and Github

Also, add to the title of the notebook, an underscore as well as your name.

Ex. '4.2_Regression_Workshop_Usama'


#Warmup

Import pandas, numpy, scipy.stats, statsmodels.api, LinearRegression from sklearn & ols from statsmodels.formula.api.

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as sc
import statsmodels.api as sm



/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


#Boston Regression

Using statsmodels and the `boston` dataset, make a regression model to predict house prices. Don't forget to add a constant (intercept) term. Note that statsmodels can take a `pd.DataFrame` as an input for `X`.

Report the $R^2$ and coefficients on each feature.

In [ ]:
from sklearn.datasets import load_boston

boston = pd.DataFrame(load_boston().data,columns = load_boston().feature_names )
y = load_boston().target


In [ ]:
# linear regression
X = sm.add_constant(boston)
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          6.72e-135
Time:                        23:10:00   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.000      26.432      46.487
CRIM          -0.1080      0.033     -3.287      0.001      -0.173      -0.043
ZN             0.0464      0.014      3.382      0.001       0.019       0.073
INDUS          0.0206      0.061      0.334      0.738      -0.100       0.141
CHAS           2.6867      0.862      3.118      0.002       0.994       4.380
NOX          -17.7666      3.820     -4.651      0.000     -25.272     -10.262
RM             3.8099      0.418      9.116      0.000       2.989       4.631
AGE            0.0007      0.013      0.052      0.958      -0.025       0.027
DIS           -1.4756      0.199     -7.398      0.000      -1.867      -1.084
RAD            0.3060      0.066      4.613      0.000       0.176       0.436
TAX           -0.0123      0.004     -3.280      0.001      -0.020      -0.005
PTRATIO       -0.9527      0.131     -7.283      0.000      -1.210      -0.696
B              0.0093      0.003      3.467      0.001       0.004       0.015
LSTAT         -0.5248      0.051    -10.347      0.000      -0.624      -0.425
==============================================================================
Omnibus:                      178.041   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              783.126
Skew:                           1.521   Prob(JB):                    8.84e-171
Kurtosis:                       8.281   Cond. No.                     1.51e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
# deleting non-related column
X = X.drop(columns=['INDUS', 'AGE'])
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     128.2
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          5.54e-137
Time:                        23:11:13   Log-Likelihood:                -1498.9
No. Observations:                 506   AIC:                             3022.
Df Residuals:                     494   BIC:                             3072.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.3411      5.067      7.171      0.000      26.385      46.298
CRIM          -0.1084      0.033     -3.307      0.001      -0.173      -0.044
ZN             0.0458      0.014      3.390      0.001       0.019       0.072
CHAS           2.7187      0.854      3.183      0.002       1.040       4.397
NOX          -17.3760      3.535     -4.915      0.000     -24.322     -10.430
RM             3.8016      0.406      9.356      0.000       3.003       4.600
DIS           -1.4927      0.186     -8.037      0.000      -1.858      -1.128
RAD            0.2996      0.063      4.726      0.000       0.175       0.424
TAX           -0.0118      0.003     -3.493      0.001      -0.018      -0.005
PTRATIO       -0.9465      0.129     -7.334      0.000      -1.200      -0.693
B              0.0093      0.003      3.475      0.001       0.004       0.015
LSTAT         -0.5226      0.047    -11.019      0.000      -0.616      -0.429
==============================================================================
Omnibus:                      178.430   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              787.785
Skew:                           1.523   Prob(JB):                    8.60e-172
Kurtosis:                       8.300   Cond. No.                     1.47e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Use polynomial features to improve your regression model in `2.1`. You can use squared and cubic features. Try to find a model that minimizes the `AIC` or `BIC` of your output table.

In [ ]:
boston = boston.drop(columns=['INDUS', 'AGE'])

In [ ]:
boston.columns

Index(['CRIM', 'ZN', 'CHAS', 'NOX', 'RM', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B',
       'LSTAT'],
      dtype='object')

In [ ]:
X = sm.add_constant(boston)
for column in boston.columns:
  X = np.concatenate([X, np.square(boston[column])[:, np.newaxis]], axis=1)
for column in boston.columns:
  X = np.concatenate([X, (boston[column]**3)[:, np.newaxis]], axis=1)

sm.OLS(y, X).fit().summary()


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  """


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.844
Model:                            OLS   Adj. R-squared:                  0.834
Method:                 Least Squares   F-statistic:                     82.78
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          5.73e-170
Time:                        23:16:33   Log-Likelihood:                -1370.0
No. Observations:                 506   AIC:                             2804.
Df Residuals:                     474   BIC:                             2939.
Df Model:                          31                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        155.1562     91.565      1.694      0.091     -24.768     335.080
x1            -0.6503      0.165     -3.944      0.000      -0.974      -0.326
x2             0.0386      0.062      0.624      0.533      -0.083       0.160
x3             0.8545      0.230      3.708      0.000       0.402       1.307
x4           379.3342    144.938      2.617      0.009      94.535     664.134
x5           -25.9062     15.864     -1.633      0.103     -57.079       5.266
x6            -5.8808      1.337     -4.400      0.000      -8.507      -3.254
x7             1.5948      0.662      2.409      0.016       0.294       2.896
x8            -0.2052      0.064     -3.230      0.001      -0.330      -0.080
x9           -11.6968     14.822     -0.789      0.430     -40.822      17.429
x10           -0.0032      0.028     -0.115      0.909      -0.057       0.051
x11           -1.8787      0.285     -6.587      0.000      -2.439      -1.318
x12            0.0125      0.005      2.367      0.018       0.002       0.023
x13           -0.0024      0.002     -1.200      0.231      -0.006       0.002
x14            0.8545      0.230      3.708      0.000       0.402       1.307
x15         -654.4390    236.214     -2.771      0.006   -1118.594    -190.283
x16            2.8412      2.529      1.123      0.262      -2.129       7.811
x17            0.8002      0.240      3.331      0.001       0.328       1.272
x18           -0.1721      0.095     -1.806      0.072      -0.359       0.015
x19            0.0005      0.000      2.975      0.003       0.000       0.001
x20            0.5154      0.864      0.597      0.551      -1.182       2.213
x21            0.0001      0.000      0.793      0.428      -0.000       0.000
x22            0.0604      0.017      3.492      0.001       0.026       0.094
x23        -7.937e-05    4.4e-05     -1.806      0.072      -0.000       7e-06
x24         2.432e-05   1.57e-05      1.552      0.121   -6.48e-06    5.51e-05
x25            0.8545      0.230      3.708      0.000       0.402       1.307
x26          342.3962    125.060      2.738      0.006      96.655     588.137
x27           -0.0581      0.132     -0.440      0.660      -0.318       0.202
x28           -0.0372      0.013     -2.820      0.005      -0.063      -0.011
x29            0.0052      0.003      1.759      0.079      -0.001       0.011
x30        -3.527e-07   1.22e-07     -2.896      0.004   -5.92e-07   -1.13e-07
x31           -0.0077      0.017     -0.462      0.644      -0.040       0.025
x32        -2.345e-07   2.15e-07     -1.091      0.276   -6.57e-07    1.88e-07
x33           -0.0007      0.000     -2.232      0.026      -0.001   -8.43e-05
==============================================================================
Omnibus:                      131.765   Durbin-Watson:                   1.269
Pr

In [ ]:
square = ['CRIM','CHAS','NOX','DIS','TAX','LSTAT']

X = sm.add_constant(boston)
for column in square:
  X = np.concatenate([X, np.square(boston[column])[:, np.newaxis]], axis=1)
for column in boston.columns:
  X = np.concatenate([X, (boston[column]**3)[:, np.newaxis]], axis=1)

sm.OLS(y, X).fit().summary()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  import sys


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.842
Model:                            OLS   Adj. R-squared:                  0.833
Method:                 Least Squares   F-statistic:                     98.06
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          1.99e-173
Time:                        23:18:01   Log-Likelihood:                -1373.7
No. Observations:                 506   AIC:                             2801.
Df Residuals:                     479   BIC:                             2915.
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         89.5333     32.942      2.718      0.007      24.805     154.261
x1            -0.6318      0.162     -3.912      0.000      -0.949      -0.314
x2            -0.0327      0.023     -1.402      0.162      -0.078       0.013
x3             0.8408      0.230      3.653      0.000       0.389       1.293
x4           338.4945    143.103      2.365      0.018      57.307     619.682
x5            -8.2346      1.420     -5.801      0.000     -11.024      -5.445
x6            -5.6611      1.279     -4.425      0.000      -8.175      -3.147
x7             0.5058      0.156      3.250      0.001       0.200       0.812
x8            -0.2338      0.062     -3.774      0.000      -0.356      -0.112
x9            -3.4070      0.848     -4.018      0.000      -5.073      -1.741
x10            0.0167      0.007      2.539      0.011       0.004       0.030
x11           -1.9154      0.279     -6.854      0.000      -2.464      -1.366
x12            0.0118      0.005      2.283      0.023       0.002       0.022
x13            0.8408      0.230      3.653      0.000       0.389       1.293
x14         -586.5136    233.194     -2.515      0.012   -1044.723    -128.304
x15            0.7731      0.230      3.368      0.001       0.322       1.224
x16            0.0005      0.000      3.457      0.001       0.000       0.001
x17            0.0613      0.017      3.602      0.000       0.028       0.095
x18        -7.311e-05   4.31e-05     -1.696      0.091      -0.000    1.16e-05
x19         5.945e-06   2.69e-06      2.206      0.028     6.5e-07    1.12e-05
x20            0.8408      0.230      3.653      0.000       0.389       1.293
x21          305.8244    123.422      2.478      0.014      63.309     548.340
x22            0.0895      0.011      8.096      0.000       0.068       0.111
x23           -0.0365      0.013     -2.878      0.004      -0.061      -0.012
x24           -0.0002      0.000     -0.738      0.461      -0.001       0.000
x25        -3.937e-07   1.18e-07     -3.323      0.001   -6.27e-07   -1.61e-07
x26            0.0028      0.001      3.098      0.002       0.001       0.004
x27        -5.946e-08   3.43e-08     -1.734      0.084   -1.27e-07    7.92e-09
x28           -0.0007      0.000     -2.291      0.022      -0.001      -0.000
==============================================================================
Omnibus:                      129.564   Durbin-Watson:                   1.278
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2751.666
Skew:                           0.529   Prob(JB):                         0.00
Kurtosis:                      14.375   Cond. No.                     1.37e+16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is

In [ ]:
cubic = ['ZN','CHAS','NOX','RM','DIS','TAX','PTRATIO','LSTAT']

In [ ]:
X = sm.add_constant(boston)

for column in square:
  X = np.concatenate([X, np.square(boston[column])[:, np.newaxis]], axis=1)
for column in cubic:
  X = np.concatenate([X, (boston[column]**3)[:, np.newaxis]], axis=1)

sm.OLS(y, X).fit().summary()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  after removing the cwd from sys.path.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:6: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.840
Model:                            OLS   Adj. R-squared:                  0.832
Method:                 Least Squares   F-statistic:                     109.6
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          5.68e-175
Time:                        23:21:58   Log-Likelihood:                -1377.4
No. Observations:                 506   AIC:                             2803.
Df Residuals:                     482   BIC:                             2904.
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         82.8548     31.523      2.628      0.009      20.914     144.795
x1            -0.4214      0.082     -5.134      0.000      -0.583      -0.260
x2            -0.0297      0.023     -1.285      0.199      -0.075       0.016
x3             0.8643      0.231      3.746      0.000       0.411       1.318
x4           375.4559    139.571      2.690      0.007     101.214     649.698
x5            -8.7609      1.409     -6.218      0.000     -11.529      -5.992
x6            -5.4620      1.263     -4.325      0.000      -7.944      -2.980
x7             0.3488      0.070      4.950      0.000       0.210       0.487
x8            -0.2211      0.059     -3.737      0.000      -0.337      -0.105
x9            -3.3483      0.838     -3.998      0.000      -4.994      -1.703
x10            0.0059      0.002      2.722      0.007       0.002       0.010
x11           -1.9123      0.280     -6.835      0.000      -2.462      -1.363
x12            0.0034      0.001      3.258      0.001       0.001       0.006
x13            0.8643      0.231      3.746      0.000       0.411       1.318
x14         -644.4612    227.735     -2.830      0.005   -1091.937    -196.985
x15            0.7383      0.228      3.239      0.001       0.290       1.186
x16            0.0005      0.000      3.412      0.001       0.000       0.001
x17            0.0607      0.017      3.575      0.000       0.027       0.094
x18         5.453e-06   2.67e-06      2.043      0.042    2.09e-07    1.07e-05
x19            0.8643      0.231      3.746      0.000       0.411       1.318
x20          335.2968    120.714      2.778      0.006      98.106     572.488
x21            0.0930      0.011      8.457      0.000       0.071       0.115
x22           -0.0344      0.013     -2.722      0.007      -0.059      -0.010
x23        -3.797e-07   1.16e-07     -3.277      0.001   -6.07e-07   -1.52e-07
x24            0.0027      0.001      3.042      0.002       0.001       0.004
x25           -0.0007      0.000     -2.275      0.023      -0.001   -9.62e-05
==============================================================================
Omnibus:                      129.062   Durbin-Watson:                   1.284
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2466.884
Skew:                           0.565   Prob(JB):                         0.00
Kurtosis:                      13.758   Cond. No.                     1.34e+16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 7.1e-14. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [ ]:
X = sm.add_constant(boston)
X = X.drop(columns=['ZN'])

for column in square:
  X = np.concatenate([X, np.square(boston[column])[:, np.newaxis]], axis=1)
for column in cubic:
  X = np.concatenate([X, (boston[column]**3)[:, np.newaxis]], axis=1)

sm.OLS(y, X).fit().summary()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  """
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:7: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  import sys


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.839
Model:                            OLS   Adj. R-squared:                  0.832
Method:                 Least Squares   F-statistic:                     114.4
Date:                Wed, 28 Apr 2021   Prob (F-statistic):          1.19e-175
Time:                        23:24:53   Log-Likelihood:                -1378.2
No. Observations:                 506   AIC:                             2802.
Df Residuals:                     483   BIC:                             2900.
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         69.5928     29.806      2.335      0.020      11.027     128.159
x1            -0.4156      0.082     -5.068      0.000      -0.577      -0.254
x2             0.8615      0.231      3.732      0.000       0.408       1.315
x3           424.0644    134.437      3.154      0.002     159.910     688.219
x4            -8.9437      1.403     -6.376      0.000     -11.700      -6.188
x5            -5.2545      1.253     -4.192      0.000      -7.717      -2.792
x6             0.3396      0.070      4.841      0.000       0.202       0.477
x7            -0.2259      0.059     -3.825      0.000      -0.342      -0.110
x8            -3.0704      0.810     -3.792      0.000      -4.661      -1.479
x9             0.0059      0.002      2.712      0.007       0.002       0.010
x10           -1.8890      0.279     -6.762      0.000      -2.438      -1.340
x11            0.0034      0.001      3.199      0.001       0.001       0.005
x12            0.8615      0.231      3.732      0.000       0.408       1.315
x13         -717.2891    220.719     -3.250      0.001   -1150.977    -283.602
x14            0.6994      0.226      3.093      0.002       0.255       1.144
x15            0.0005      0.000      3.503      0.001       0.000       0.001
x16            0.0597      0.017      3.516      0.000       0.026       0.093
x17         2.713e-06   1.61e-06      1.690      0.092   -4.42e-07    5.87e-06
x18            0.8615      0.231      3.732      0.000       0.408       1.315
x19          371.3765    117.483      3.161      0.002     140.536     602.217
x20            0.0944      0.011      8.626      0.000       0.073       0.116
x21           -0.0326      0.013     -2.596      0.010      -0.057      -0.008
x22        -3.897e-07   1.16e-07     -3.368      0.001   -6.17e-07   -1.62e-07
x23            0.0024      0.001      2.823      0.005       0.001       0.004
x24           -0.0007      0.000     -2.233      0.026      -0.001   -8.32e-05
==============================================================================
Omnibus:                      131.667   Durbin-Watson:                   1.276
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2505.865
Skew:                           0.594   Prob(JB):                         0.00
Kurtosis:                      13.837   Cond. No.                     5.14e+25
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 4.85e-33. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

After applying squared and cubic forms, AIC is minimized to 2802 from 3026 and BIC is minimized to 2900 from 3085.

Features using squared form and cubic form are listed below.

In [ ]:
print(square)
print(cubic)

['CRIM', 'CHAS', 'NOX', 'DIS', 'TAX', 'LSTAT']
['ZN', 'CHAS', 'NOX', 'RM', 'DIS', 'TAX', 'PTRATIO', 'LSTAT']


#Regressive Tennis T-Test

Use a linear regression and statsmodels to run a t-test on whether Federer scores more points than his opponents in the `tennis.csv` dataset.


Also answer the following: should your regression include a constant term? Why or why not? How would it change the interpretation of your coefficient and p-value?

In [ ]:
from google.colab import files
uploaded = files.upload()

In [55]:
df = pd.read_csv('tennis.csv')
df.head()

year           tournament  ... player2 total points total    win
0  1998   Basel, Switzerland  ...                      103.0  False
1  1998     Toulouse, France  ...                      106.0   True
2  1998     Toulouse, France  ...                      148.0   True
3  1998     Toulouse, France  ...                      130.0  False
4  1998  Geneva, Switzerland  ...                        NaN  False

[5 rows x 71 columns]

In [57]:
df = df.dropna()
df['win'] = 0
df.loc[(df['winner']==df['player1 name']),['win']] = 1

In [59]:
sc.ttest_1samp(df['win'], 0.5)

Ttest_1sampResult(statistic=26.245156533243442, pvalue=2.6923903460915165e-116)

By running a t-test with pvalue=2.6923903460915165e-116, null hypothesis is rejected and Federer scores more points than his opponents.

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1016 entries, 0 to 1177
Data columns (total 71 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   year                                   1016 non-null   int64  
 1   tournament                             1016 non-null   object 
 2   start date                             1016 non-null   object 
 3   type                                   1016 non-null   object 
 4   surface                                1016 non-null   object 
 5   draw                                   1016 non-null   object 
 6   atp points                             1016 non-null   object 
 7   atp ranking                            1016 non-null   float64
 8   tournament prize money                 1016 non-null   object 
 9   round                                  1016 non-null   object 
 10  opponent                               1016 non-null   object 
 11  rank

In [64]:
# get string columns
obj_col = []
for column in df.columns:
  if df[column].dtype == 'object':
    obj_col.append(column)


In [ ]:
df_r = df.drop(columns=obj_col)
df_r.info()

In [68]:
# regression without intercept
X = df_r.drop(columns=['win'])
y = df_r['win']

sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    win   R-squared (uncentered):                   0.938
Model:                            OLS   Adj. R-squared (uncentered):              0.936
Method:                 Least Squares   F-statistic:                              678.5
Date:                Thu, 29 Apr 2021   Prob (F-statistic):                        0.00
Time:                        02:16:47   Log-Likelihood:                          69.479
No. Observations:                1016   AIC:                                     -94.96
Df Residuals:                     994   BIC:                                      13.36
Df Model:                          22                                                  
Covariance Type:            nonrobust                                                  
=========================================================================================================
                                            coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
year                                      0.0003   1.32e-05     24.421      0.000       0.000       0.000
atp ranking                              -0.0002      0.000     -1.285      0.199      -0.000    8.66e-05
time                                     -0.0007      0.001     -1.032      0.302      -0.002       0.001
player1 aces                             -0.0022      0.002     -1.037      0.300      -0.006       0.002
player1 double faults                     0.0062      0.005      1.246      0.213      -0.004       0.016
player1 1st serves in                     0.0011      0.000      2.268      0.024       0.000       0.002
player1 1st serves total                  0.0009      0.000      2.553      0.011       0.000       0.002
player1 1st serve points won             -0.0004      0.002     -0.286      0.775      -0.003       0.003
player1 1st serve points total            0.0011      0.000      2.268      0.024       0.000       0.002
player1 2nd serve points won              0.0046      0.001      4.045      0.000       0.002       0.007
player1 2nd serve points total           -0.0001      0.000     -0.358      0.720      -0.001       0.001
player1 break points won                  0.0274      0.006      4.812      0.000       0.016       0.039
player1 break points total               -0.0124      0.003     -4.646      0.000      -0.018      -0.007
player1 service games played              0.0021      0.007      0.303      0.762      -0.012       0.016
player1 1st serve return points won       0.0010      0.001      0.810      0.418      -0.001       0.003
player1 1st serve return points total     0.0006      0.000      1.358      0.175      -0.000       0.001
player1 2nd serve return points won       0.0015      0.001      1.486      0.137      -0.000       0.004
player1 2nd serve return points total     0.0005      0.000      2.055      0.040    2.34e-05       0.001
player1 break points converted won        0.0011      0.008      0.144      0.885      -0.014       0.016
player1 break points converted total     -0.0002      0.002     -0.090      0.929      -0.005       0.005
player1 return games played              -0.0127      0.007     -1.804      0.072      -0.026       0.001
player1 total service points won          0.0041      0.001      2.842      0.005       0.001       0.007
player1 total service points total        0.0009      0.000      2.553      0.011       0.000       0.002
player1 total return points won           0.0025      0.001      1.927      0.054   -4.67e-05       0.005
player1 total return points total         0.0011      0.001      2.139      0.033    9.31e-05       0.002
player1 total po

In [69]:
# regression with intercept
X = sm.add_constant(X)

sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    win   R-squared:                       0.657
Model:                            OLS   Adj. R-squared:                  0.650
Method:                 Least Squares   F-statistic:                     86.51
Date:                Thu, 29 Apr 2021   Prob (F-statistic):          6.54e-213
Time:                        02:17:44   Log-Likelihood:                 69.504
No. Observations:                1016   AIC:                            -93.01
Df Residuals:                     993   BIC:                             20.24
Df Model:                          22                                         
Covariance Type:            nonrobust                                         
=========================================================================================================
                                            coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
const                                    -1.0719      4.891     -0.219      0.827     -10.670       8.526
year                                      0.0009      0.002      0.351      0.725      -0.004       0.006
atp ranking                              -0.0002      0.000     -1.188      0.235      -0.000       0.000
time                                     -0.0007      0.001     -1.031      0.303      -0.002       0.001
player1 aces                             -0.0022      0.002     -1.045      0.296      -0.006       0.002
player1 double faults                     0.0064      0.005      1.264      0.206      -0.004       0.016
player1 1st serves in                     0.0011      0.000      2.269      0.023       0.000       0.002
player1 1st serves total                  0.0009      0.000      2.561      0.011       0.000       0.002
player1 1st serve points won             -0.0005      0.002     -0.299      0.765      -0.003       0.003
player1 1st serve points total            0.0011      0.000      2.269      0.023       0.000       0.002
player1 2nd serve points won              0.0046      0.001      4.049      0.000       0.002       0.007
player1 2nd serve points total           -0.0001      0.000     -0.343      0.732      -0.001       0.001
player1 break points won                  0.0274      0.006      4.806      0.000       0.016       0.039
player1 break points total               -0.0123      0.003     -4.640      0.000      -0.018      -0.007
player1 service games played              0.0020      0.007      0.293      0.770      -0.012       0.016
player1 1st serve return points won       0.0010      0.001      0.809      0.419      -0.001       0.003
player1 1st serve return points total     0.0006      0.000      1.356      0.175      -0.000       0.001
player1 2nd serve return points won       0.0015      0.001      1.485      0.138      -0.000       0.004
player1 2nd serve return points total     0.0005      0.000      2.065      0.039    2.62e-05       0.001
player1 break points converted won        0.0011      0.008      0.141      0.888      -0.014       0.016
player1 break points converted total     -0.0002      0.002     -0.097      0.923      -0.005       0.005
player1 return games played              -0.0126      0.007     -1.800      0.072      -0.026       0.001
player1 total service points won          0.0041      0.001      2.836      0.005       0.001       0.007
player1 total service points total        0.0009      0.000      2.561      0.011       0.000       0.002
player1 total return points won           0.0025      0.001      1.924      0.055   -5.06e-05       0.005
player1 total return points total         0.0011      0.001      2.148      0.032    9.77e-05       0.002
player1 t

Regression model without intercept performs better with higher R^2 value.

#College Admissions

Using the `college.csv` dataset, answer the following:

1. Do private schools see more admissions overall? Hint: use a binary explanatory variable for `Private`

2. Is the relation between `Top10perc` and `Top25perc` best fit using a model with only one variable, or one variable and a polynomial of degree 2? Is a constant term useful? How would you select for the best of these model specifications?

In [ ]:
from google.colab import files
uploaded = files.upload()

In [20]:
import io
df = pd.read_csv(io.BytesIO(uploaded['college.csv']))
df = df.dropna()

In [21]:
df.head()

Unnamed: 0 Private  Apps  ...  perc.alumni  Expend  Grad.Rate
0  Abilene Christian University     Yes  1660  ...           12    7041         60
1            Adelphi University     Yes  2186  ...           16   10527         56
2                Adrian College     Yes  1428  ...           30    8735         54
3           Agnes Scott College     Yes   417  ...           37   19016         59
4     Alaska Pacific University     Yes   193  ...            2   10922         15

[5 rows x 19 columns]

In [22]:
df_p = df.loc[:,['Private']]
df_p = df_p.dropna()
df_p.loc[df_p['Private']=='Yes',['Private']] = 1
df_p.loc[df_p['Private']!=1,['Private']] = 0


In [23]:
sc.ttest_1samp(df_p['Private'], 0.5)

Ttest_1sampResult(statistic=14.206379631727932, pvalue=6.967434784281823e-41)

As t-test result shown, null hypothesis can be rejected. Thus, private schools see more admissions overall.

In [24]:
X = df['Top10perc']
y = df['Top25perc']

# linear no intercept
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              Top25perc   R-squared (uncentered):                   0.911
Model:                            OLS   Adj. R-squared (uncentered):              0.911
Method:                 Least Squares   F-statistic:                              7971.
Date:                Thu, 29 Apr 2021   Prob (F-statistic):                        0.00
Time:                        01:07:50   Log-Likelihood:                         -3332.4
No. Observations:                 777   AIC:                                      6667.
Df Residuals:                     776   BIC:                                      6671.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Top10perc      1.7275      0.019     89.281      0.000       1.690       1.766
==============================================================================
Omnibus:                      290.739   Durbin-Watson:                   1.499
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1310.349
Skew:                          -1.681   Prob(JB):                    2.89e-285
Kurtosis:                       8.401   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [25]:
# linear with intercept
X = sm.add_constant(X)
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Top25perc   R-squared:                       0.796
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     3018.
Date:                Thu, 29 Apr 2021   Prob (F-statistic):          1.87e-269
Time:                        01:07:58   Log-Likelihood:                -2805.2
No. Observations:                 777   AIC:                             5614.
Df Residuals:                     775   BIC:                             5624.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         28.1984      0.596     47.281      0.000      27.028      29.369
Top10perc      1.0014      0.018     54.933      0.000       0.966       1.037
==============================================================================
Omnibus:                       45.943   Durbin-Watson:                   1.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               85.462
Skew:                           0.401   Prob(JB):                     2.77e-19
Kurtosis:                       4.413   Cond. No.                         60.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [27]:
# polynomial no intercept
X = df.loc[:,['Top10perc']]
X = np.concatenate([X, (df['Top10perc']**2)[:, np.newaxis]], axis=1)
 
sm.OLS(y, X).fit().summary()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  This is separate from the ipykernel package so we can avoid doing imports until


<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              Top25perc   R-squared (uncentered):                   0.975
Model:                            OLS   Adj. R-squared (uncentered):              0.975
Method:                 Least Squares   F-statistic:                          1.515e+04
Date:                Thu, 29 Apr 2021   Prob (F-statistic):                        0.00
Time:                        01:11:06   Log-Likelihood:                         -2839.4
No. Observations:                 777   AIC:                                      5683.
Df Residuals:                     775   BIC:                                      5692.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1             2.7198      0.025    110.830      0.000       2.672       2.768
x2            -0.0195      0.000    -44.515      0.000      -0.020      -0.019
==============================================================================
Omnibus:                      129.648   Durbin-Watson:                   1.927
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              287.613
Skew:                           0.919   Prob(JB):                     3.51e-63
Kurtosis:                       5.347   Cond. No.                         134.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [28]:
# polynomial with intercept
X = sm.add_constant(X)

sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Top25perc   R-squared:                       0.852
Model:                            OLS   Adj. R-squared:                  0.851
Method:                 Least Squares   F-statistic:                     2220.
Date:                Thu, 29 Apr 2021   Prob (F-statistic):          2.50e-321
Time:                        01:11:48   Log-Likelihood:                -2681.0
No. Observations:                 777   AIC:                             5368.
Df Residuals:                     774   BIC:                             5382.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         16.6786      0.845     19.738      0.000      15.020      18.337
x1             1.8140      0.050     36.229      0.000       1.716       1.912
x2            -0.0102      0.001    -17.072      0.000      -0.011      -0.009
==============================================================================
Omnibus:                      177.730   Durbin-Watson:                   1.960
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              570.823
Skew:                           1.089   Prob(JB):                    1.12e-124
Kurtosis:                       6.590   Cond. No.                     5.65e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.65e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Among four diffrent models, polynomial without intercept performs best with highest R^2 value(0.975).

# Auto Multiple Regression

Using the `auto.csv` dataset, perform a multiple linear regression with `mpg` as the response and all other variables except name in `auto` as the predictors. Comment on the output:

i. Is there a relationship between the predictors and the response?

ii. Which predictors appear to have a statistically significant relationship to the response?

iii. What does the coefficient for the year variable suggest?

In [ ]:
from google.colab import files
uploaded = files.upload()

In [44]:
df = pd.read_csv('auto.csv', delim_whitespace = True)
df = df.dropna()
df.head()

mpg  cylinders  displacement  ...  year  origin                       name
1  18.0        8.0         307.0  ...  70.0     1.0  chevrolet chevelle malibu
2  15.0        8.0         350.0  ...  70.0     1.0          buick skylark 320
3  18.0        8.0         318.0  ...  70.0     1.0         plymouth satellite
4  16.0        8.0         304.0  ...  70.0     1.0              amc rebel sst
5  17.0        8.0         302.0  ...  70.0     1.0                ford torino

[5 rows x 9 columns]

In [ ]:
df_1 = df.drop(columns=['name'])


In [49]:
df_1['horsepower'].unique()

array(['130.0', '165.0', '150.0', '140.0', '198.0', '220.0', '215.0',
       '225.0', '190.0', '170.0', '160.0', '95.00', '97.00', '85.00',
       '88.00', '46.00', '87.00', '90.00', '113.0', '200.0', '210.0',
       '193.0', '?', '100.0', '105.0', '175.0', '153.0', '180.0', '110.0',
       '72.00', '86.00', '70.00', '76.00', '65.00', '69.00', '60.00',
       '80.00', '54.00', '208.0', '155.0', '112.0', '92.00', '145.0',
       '137.0', '158.0', '167.0', '94.00', '107.0', '230.0', '49.00',
       '75.00', '91.00', '122.0', '67.00', '83.00', '78.00', '52.00',
       '61.00', '93.00', '148.0', '129.0', '96.00', '71.00', '98.00',
       '115.0', '53.00', '81.00', '79.00', '120.0', '152.0', '102.0',
       '108.0', '68.00', '58.00', '149.0', '89.00', '63.00', '48.00',
       '66.00', '139.0', '103.0', '125.0', '133.0', '138.0', '135.0',
       '142.0', '77.00', '62.00', '132.0', '84.00', '64.00', '74.00',
       '116.0', '82.00'], dtype=object)

In [51]:
df_1 = df_1[df_1['horsepower']!='?']
df_1 = df_1.astype('float')
df_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 1 to 397
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    float64
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    float64
 5   acceleration  392 non-null    float64
 6   year          392 non-null    float64
 7   origin        392 non-null    float64
dtypes: float64(8)
memory usage: 27.6 KB


In [52]:
# OLS regression
X = df_1.drop(columns=['mpg'])
y = df_1['mpg']

sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    mpg   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.981
Method:                 Least Squares   F-statistic:                              2933.
Date:                Thu, 29 Apr 2021   Prob (F-statistic):                        0.00
Time:                        01:35:52   Log-Likelihood:                         -1030.4
No. Observations:                 392   AIC:                                      2075.
Df Residuals:                     385   BIC:                                      2103.
Df Model:                           7                                                  
Covariance Type:            nonrobust                                                  
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
cylinders       -0.7055      0.323     -2.181      0.030      -1.341      -0.070
displacement     0.0219      0.008      2.871      0.004       0.007       0.037
horsepower      -0.0396      0.013     -3.147      0.002      -0.064      -0.015
weight          -0.0060      0.001     -9.258      0.000      -0.007      -0.005
acceleration    -0.0796      0.090     -0.881      0.379      -0.257       0.098
year             0.5836      0.024     24.150      0.000       0.536       0.631
origin           1.3100      0.281      4.663      0.000       0.758       1.862
==============================================================================
Omnibus:                       46.837   Durbin-Watson:                   1.283
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               90.015
Skew:                           0.681   Prob(JB):                     2.84e-20
Kurtosis:                       4.912   Cond. No.                     6.00e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large,  6e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

i. Is there a relationship between the predictors and the response?
With large value of R^2  and F-stat, there exists relationship between the predictors and the response. 

ii. Which predictors appear to have a statistically significant relationship to the response?
Cylinders, displacement, horsepower, weight, year, and origin ave a statistically significant relationship to the response.

iii. What does the coefficient for the year variable suggest?
0.5836